developer-hub/src/reference_docs/safe_defensive_programming.rs

```diff
@@ -1 +1,615 @@
+//! As our runtime should _never_ panic; this includes carefully handling [`Result`]/[`Option`]
+//! types, eliminating the possibility of integer overflows, converting between number types, or
+//! even handling floating point usage with fixed point arithmetic to mitigate issues that come with
+//! floating point calculations.
 //!
+//!
+//! ## Defensive Programming
+//!
+//! Defensive programming is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
+//! graceful manner.
+//!
+//! The Polkadot SDK is both built to reflect these principles, and to be used accordingly.
+//!
+//! ## General Practices
+//!
+//! When developing within the context of the a runtime, there is one golden rule:
+//!
+//! ***DO NOT PANIC!***
+//!
+//! Most of the time, unless you wish for your node to be intentionally brought down - panicking is
+//! something that your runtime should feverishly protect against. This includes the following
+//! considerations:
```

Comment:
There are some panics in Substrate, i.e. https://github.com/paritytech/polkadot-sdk/blob/126f64a91ee0c0333f368fe67bb00d56d68ecb34/substrate/frame/aura/src/lib.rs#L147

Reply:
Doesn't mean it is best practice :P

Reply:
My impression is that panicking should only occur if in the event that it would actually be better (in a rare case) to shut down rather than continue operating.  This is probably only for 'critical' components, i.e., finality, authoring, or consensus related, but I could be wrong here.

The aura example is a good one, I'll link to it from within and rephrase accordingly.